### PR TITLE
MINOR: update KTable JavaDoc

### PR DESCRIPTION
**Pull request by mjsax**

### CODE REVIEW

**Comment by mjsax** [comment]
@dguy @enothereska @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/kstream/GlobalKTable.java [hunk 1]
```diff
@@ -19,23 +19,28 @@
...
- * A {@link GlobalKTable} is fully replicated per {@link KafkaStreams} instance. Every partition of the underlying topic
- * is consumed by each {@link GlobalKTable}, such that the full set of data is available in every {@link KafkaStreams} instance.
- * This provides the ability to perform joins with {@link KStream}, and {@link KTable},
- * without having to repartition the input streams. All joins with the {@link GlobalKTable} require that a {@link KeyValueMapper}
- * is provided that can map from the (key, value) of the left hand side to the key of the right hand side {@link GlobalKTable}
+ * In contrast to a {@link KTable} that is partitioned over all {@link KafkaStreams} instances, a {@link GlobalKTable}
+ * is fully replicated per {@link KafkaStreams} instance.
+ * Every partition of the underlying topic is consumed by each {@link GlobalKTable}, such that the full set of data is
+ * available in every {@link KafkaStreams} instance.
+ * This provides the ability to perform joins with {@link KStream} and {@link KTable}, without having to repartition the
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/GlobalKTable.java]: We should actually remove the reference to `KTable` as we don't currently support it

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 2]
```diff
@@ -17,378 +17,831 @@
...
 import org.apache.kafka.streams.processor.StreamPartitioner;
-import org.apache.kafka.clients.producer.internals.DefaultPartitioner;
+import org.apache.kafka.streams.state.ReadOnlyKeyValueStore;
 
 /**
  * {@link KTable} is an abstraction of a <i>changelog stream</i> from a primary-keyed table.
- * Each record in this stream is an update on the primary-keyed table with the record key as the primary key.
+ * Each record in this changelog stream is an update on the primary-keyed table with the record key as the primary key.
  * <p>
  * A {@link KTable} is either defined from one or multiple Kafka topics that are consumed message by message or
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: remove "or multiple"
Unless there is something i'm missing a `KTable` is created from exactly one topic unless it is the result of an aggregation

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 3]
```diff
@@ -17,378 +17,831 @@
...
     /**
-     * Print the elements of this stream to {@code System.out}
-     * @param keySerde key serde used to send key-value pairs,
-     *                 if not specified the default serde defined in the configs will be used
-     * @param valSerde value serde used to send key-value pairs,
-     *                 if not specified the default serde defined in the configs will be used
-     *
-     * Implementors will need to override toString for keys and values that are not of
-     * type String, Integer etc to get meaningful information.
+     * Print the update records of this {@link KTable} to {@code System.out}.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:179]: Print the updated...

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 4]
```diff
@@ -17,378 +17,831 @@
...
+     * This function will use the generated name of the parent processor node to label the key/value pairs printed out
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: remove out

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 5]
```diff
@@ -17,378 +17,831 @@
...
+     * to the console.
+     * <p>
+     * The provided serde will be used to deserialize the key or value in case the type is {@code byte[]} before calling
+     * {@code toString()} on the deserialized object.
+     * <p>
+     * Implementors will need to override {@code toString()} for keys and values that are not of type {@link String},
+     * {@link Integer} etc. to get meaningful information.
+     * <p>
+     * Note that {@code print()} is not applied to the internal state store and only called for each new {@link KTable}
+     * update record.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:157]: updated

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 6]
```diff
@@ -17,378 +17,831 @@
...
+               final Serde<V> valSerde);
 
     /**
-     * Print the elements of this stream to System.out
-     *
-     * @param keySerde key serde used to send key-value pairs,
-     *                 if not specified the default serde defined in the configs will be used
-     * @param valSerde value serde used to send key-value pairs,
-     *                 if not specified the default serde defined in the configs will be used
+     * Print the update records of this {@link KTable} to {@code System.out}.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: same as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 7]
```diff
@@ -17,378 +17,831 @@
...
      */
-    void print(Serde<K> keySerde, Serde<V> valSerde, String streamName);
+    void print(final Serde<K> keySerde,
+               final Serde<V> valSerde,
+               final String streamName);
 
     /**
-     * Write the elements of this stream to a file at the given path using default serializers and deserializers.
-     * @param filePath name of file to write to
+     * Write the update records of this {@link KTable} to a file at the given path.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:220]: updated

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 8]
```diff
@@ -17,378 +17,831 @@
...
+     * This function will use the generated name of the parent processor node to label the key/value pairs printed out
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: written to the file

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 9]
```diff
@@ -17,378 +17,831 @@
...
+     * @param filePath name of file to write to
      */
-    void writeAsText(String filePath);
+    void writeAsText(final String filePath);
 
     /**
-     * Write the elements of this stream to a file at the given path.
-     *
-     * @param filePath name of file to write to
+     * Write the update records of this {@link KTable} to a file at the given path.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:238]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 10]
```diff
@@ -17,378 +17,831 @@
...
-     * Implementors will need to override {@code toString} for keys and values that are not of
-     * type {@link String}, {@link Integer} etc. to get meaningful information.
      */
-    void writeAsText(String filePath, String streamName);
+    void writeAsText(final String filePath,
+                     final String streamName);
 
     /**
-     * Write the elements of this stream to a file at the given path.
+     * Write the update records of this {@link KTable} to a file at the given path.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:257]: As above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 11]
```diff
@@ -17,378 +17,831 @@
...
+     * @param keySerde key serde used to deserialize key if type is {@code byte[]},
+     * @param valSerde value serde used to deserialize value if type is {@code byte[]},
      */
-    void  writeAsText(String filePath, Serde<K> keySerde, Serde<V> valSerde);
+    void  writeAsText(final String filePath,
+                      final Serde<K> keySerde,
+                      final Serde<V> valSerde);
 
     /**
+     * Write the update records of this {@link KTable} to a file at the given path.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: As above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 12]
```diff
@@ -17,378 +17,831 @@
...
 
     /**
-     * Materialize this stream to a topic, also creates a new instance of {@link KTable} from the topic
-     * using default serializers and deserializers and producer's {@link DefaultPartitioner}.
-     * This is equivalent to calling {@link #to(String)} and {@link org.apache.kafka.streams.kstream.KStreamBuilder#table(String, String)}.
-     * The resulting {@link KTable} will be materialized in a local state
-     * store with the given store name. Also a changelog topic named "${applicationId}-${storeName}-changelog"
-     * will be automatically created in Kafka for failure recovery, where "applicationID"
-     * is specified by the user in {@link org.apache.kafka.streams.StreamsConfig}.
+     * Perform an action on each update record of this {@link KTable}.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:302]: updated

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 13]
```diff
@@ -17,378 +17,831 @@
...
+     * Note that this is a terminal operation that returns void.
+     * <p>
+     * Note that {@code foreach()} is not applied to the internal state store and only called for each new
+     * {@link KTable} update record.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:306]: new or updated record?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 14]
```diff
@@ -17,378 +17,831 @@
...
-     * @param keySerde     key serde used to send key-value pairs,
-     *                     if not specified the default key serde defined in the configuration will be used
-     * @param valSerde     value serde used to send key-value pairs,
-     *                     if not specified the default value serde defined in the configuration will be used
-     * @param topic        the topic name
-     * @param storeName    the state store name used for this KTable
-     * @return a new {@link KTable} that contains the exact same records as this {@link KTable}
+     * Convert this changelog stream to a {@link KStream} using the given {@link KeyValueMapper} to select the new key.
+     * <p>
+     * For example, you can computes the new key as the length of the value string.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: compute

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 15]
```diff
@@ -17,378 +17,831 @@
...
-    void to(Serde<K> keySerde, Serde<V> valSerde, StreamPartitioner<? super K, ? super V> partitioner, String topic);
+    void to(final String topic);
 
     /**
-     * Convert this stream to a new instance of {@link KStream}.
-     *
-     * @return a {@link KStream} that contains the same records as this {@link KTable};
-     *         the records are no longer treated as updates on a primary-keyed table,
-     *         but rather as normal key-value pairs in a record stream
+     * Materialize this changelog stream to a topic and creates a new {@link KTable} from the topic using default
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: This doesn't create a new `KTable`. It just writes the data to the topic

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 16]
```diff
@@ -17,378 +17,831 @@
...
+    void to(final StreamPartitioner<? super K, ? super V> partitioner,
+            final String topic);
 
     /**
-     *  Convert this stream to a new instance of {@link KStream} using the given {@link KeyValueMapper} to select
-     *  the new key.
-     *
-     * @param mapper  @param mapper  the instance of {@link KeyValueMapper}
-     * @param <K1> the new key type
+     * Materialize this changelog stream to a topic and creates a new {@link KTable} from the topic.
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: As above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 17]
```diff
@@ -17,378 +17,831 @@
...
-     * Combine values of this stream with another {@link KTable} stream's elements of the same key using Inner Join.
-     *
-     * @param other         the instance of {@link KTable} joined with this stream
-     * @param joiner        the instance of {@link ValueJoiner}
-     * @param <V1>          the value type of the other stream
-     * @param <R>           the value type of the new stream
-     *
-     * @return a {@link KTable} that contains join-records for each key and values computed by the given {@link ValueJoiner},
-     *         one for each matched record-pair with the same key
+     * Materialize this changelog stream to a topic and creates a new {@link KTable} from the topic using a customizable
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: As above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 18]
```diff
@@ -17,378 +17,831 @@
...
+     * The {@link KeyValueMapper} selects a new key and value (with should both have unmodified type).
+     * If the new record key is {@code null} the record will not be included in the resulting {@link KGroupedTable}
+     * <p>
+     * Because a new key is selected, an internal repartitioning topic will be created in Kafka.
+     * This topic will be named "${applicationId}-XXX-repartition", where "applicationId" is user-specified in
+     * {@link  StreamsConfig} via parameter {@link StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "XXX" is
+     * an internal generated name, and "-repartition" is a fixed suffix.
+     * You can retrieve all generated internal topic names via {@link KafkaStreams#toString()}.
+     * <p>
+     * All data of this {@link KTable} will be redistributed through the repartitioning topic by writing all update
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: updated records to it, and re-reading all updated records from it,

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 19]
```diff
@@ -17,378 +17,831 @@
...
+     * The {@link KeyValueMapper} selects a new key and value (both with potentially different type).
+     * If the new record key is {@code null} the record will not be included in the resulting {@link KGroupedTable}
+     * <p>
+     * Because a new key is selected, an internal repartitioning topic will be created in Kafka.
+     * This topic will be named "${applicationId}-XXX-repartition", where "applicationId" is user-specified in
+     * {@link  StreamsConfig} via parameter {@link StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "XXX" is
+     * an internal generated name, and "-repartition" is a fixed suffix.
+     * You can retrieve all generated internal topic names via {@link KafkaStreams#toString()}.
+     * <p>
+     * All data of this {@link KTable} will be redistributed through the repartitioning topic by writing all update
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 20]
```diff
@@ -17,378 +17,831 @@
...
-     * @param <K1>          the key type of the {@link KGroupedTable}
-     * @param <V1>          the value type of the {@link KGroupedTable}
-     *
-     * @return a {@link KGroupedTable} that contains the re-partitioned records of this {@link KTable}
+     * Join records of this {@link KTable} with another {@link KTable}'s records using non-windowed inner equi join.
+     * The join is a primary key join with join attribute {@code thisKTable.key == otherKTable.key}.
+     * The result is an ever updating {@link KTable} that represents the <em>current</em> (i.e., processing time) result
+     * of the join.
+     * <p>
+     * The join is computed by (1) updating the internal state if one {@link KTable} and (2) performing a lookup for a
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: if -> of ?
also add a , before and

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 21]
```diff
@@ -17,378 +17,831 @@
...
+     * matching record in the <em>current</em> (i.e., processing time) internal state of the other {@link KTable}.
+     * This happens in a symmetric way, i.e., for each update of either {@code this} or the {@code other} input
+     * {@link KTable} the result gets updated.
+     * <p>
+     * For each {@link KTable} record that finds an corresponding record in the other {@link KTable} the provided
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: an -> a

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 22]
```diff
@@ -17,378 +17,831 @@
...
+     * internal repartitioning topic in Kafka and write and re-read the data via this topic before the actual join.
+     * The repartitioning topic will be named "${applicationId}-XXX-repartition", where "applicationId" is
+     * user-specified in {@link StreamsConfig} via parameter
+     * {@link StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "XXX" is an internal generated name, and
+     * "-repartition" is a fixed suffix.
+     * You can retrieve all generated internal topic names via {@link KafkaStreams#toString()}.
+     * <p>
+     * Repartitioning can happen both input {@link KTable}s.
+     * For this case, all data of a {@link KTable} will be redistributed through the repartitioning topic by writing all
+     * update records to and rereading all update records from it, such that the join input {@link KTable} is
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:642]: as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java [hunk 23]
```diff
@@ -17,378 +17,831 @@
...
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:642]: as above

**Review by dguy — CHANGES_REQUESTED** [review]
Some typos/grammar changes

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java (hunk 3 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:179]: Was intended as is -- records are updates, so its an "update record" (or update-record). "updated" makes sense, too, though. Not sure what the better wording is...

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java (hunk 15 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: c&p error from `through`...

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java (hunk 20 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java]: Why a comma?
To my knowledge, for two things that are listed there is no comma: "A and B"
Only for three or more, there are command: "A, B, C, and D"

Correct my if I am wrong.

**Comment by dguy** [comment]
Fair enough

**Comment by mjsax** [comment]
Addressed @dguy comments. (Not the part about "update records" -- but we can discuss if we want to reword).

Also added `KGroupedTable` and `GlobalKTable`.

Some minor refactoring in `KStream` and `KGroupedStream` (always discovering small issues along the way...)

Call for review @eno @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java (hunk 13 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:306]: I am OK with update records.

**Comment by mjsax** [comment]
Rebased to resolve conflict.

File: streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java (hunk 3 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KTable.java:179]: I see what you mean. I just think "updated records" sounds more natural. But i'm not overly concerned

**Review by dguy — APPROVED** [review]

**Comment by guozhangwang** [comment]
Merged to trunk and 0.10.2.

**Pull request closed without merging**